libraries/AC_SafeRTL/AC_SafeRTL.h

```diff
@@ -0,0 +1,89 @@
+#pragma once
+
+#include <AP_Buffer/AP_Buffer.h>
+#include <DataFlash/DataFlash.h>
+#include <AP_HAL/AP_HAL.h>
+#include <AP_Math/AP_Math.h>
+
+#include <bitset>
+#include <vector>
+
+#define SMALL_FLOAT  0.0000001f
+
+#define POSITION_DELTA 2.0f // how many meters to move before appending a new position to return_path
+#define PRUNING_DELTA (POSITION_DELTA * 0.99) // XXX must be smaller than position_delta! How many meters apart must two points be, such that we can assume that there is no obstacle between those points
+#define RDP_EPSILON (POSITION_DELTA * 0.5)
+#define MAX_PATH_LEN 100 // the amount of memory used by safe RTL will be slightly higher than 3*8*MAX_PATH_LEN bytes. Increasing this number will improve path pruning, but will use more memory, and running a path cleanup will take longer. No longer than 255.
+#define RDP_STACK_LEN 64 // the amount of memory to be allocated for the RDP algorithm to write its to do list.
+// XXX A number too small for RDP_STACK_LEN can cause a buffer overflow! The number to put here is int((s/2-1)+min(s/2, MAX_PATH_LEN-s)), where s = pow(2, floor(log(MAX_PATH_LEN)/log(2)))
+// To avoid this annoying math, a good-enough overestimate is ciel(MAX_PATH_LEN*2./3.)
+#define RDP_TIME 200
+#define LOOP_TIME 300
+
+#define HYPOT(a,b) (a-b).length()
+
+class SafeRTL_Path {
+    // points are stored in meters from EKF origin in NED
+    Vector3f path[MAX_PATH_LEN];
+    uint8_t _last_index;
+public:
+    bool accepting_new_points; // false means that any call to append_if_far_enough() will fail. This should be unset when entering SafeRTL mode, and set when exiting.
+    SafeRTL_Path(bool);
+    void append_if_far_enough(Vector3f);
+    bool routine_cleanup();
+    Vector3f* thorough_cleanup();
+    Vector3f get(int index) { return path[index]; }
+    bool pop_point(Vector3f&);
+    void reset_path(Vector3f);
+    bool cleanup_ready() { return _pruning_complete && _simplification_complete; }
+    bool is_active() { return _active; }
```

Comment:
Best to add "const" (i.e. "bool is_active() const") to these functions that don't modify the state of the object.